ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt

```diff
@@ -111,102 +110,93 @@ class SingleSignOnProfileVerifier(private val response: Node,
-            verifyBearerSubjectConfirmations(it)
-
-            // todo - We can't throw an exception here because the spec says there could be
-            // assertions without bearer
-            verifyAssertionsHaveAuthnStmts(it)
-            verifySingleLogoutIncludesSessionIndex(it)
-
-            // Assuming the AudienceRestriction is under Conditions
-            verifyBearerAssertionsContainAudienceRestriction(it)
-        }
-    }
-
-    private fun verifyBearerAssertionsContainAudienceRestriction(assertion: Node) {
-        if (assertion.children("Conditions").isNotEmpty()) {
-            val audience = assertion.children("Conditions")
-                    .firstOrNull()
-                    ?.children("AudienceRestriction")
-                    ?.firstOrNull()
-                    ?.children("Audience")
-                    ?.firstOrNull()
-                    ?.textContent
-
-            if (audience != SP_ISSUER)
-                throw SAMLComplianceException.createWithPropertyMessage(SAMLProfiles_4_1_4_2_k,
-                        property = "Audience",
-                        actual = audience,
-                        expected = SP_ISSUER,
-                        node = response)
-        }
-    }
-
-    private fun verifySingleLogoutIncludesSessionIndex(assertion: Node) {
-        if (idpMetadata.descriptor == null) return
-
-        if (idpMetadata.descriptor?.singleLogoutServices?.isNotEmpty() == true) return
-
-        assertion.children("AuthnStatement").forEach {
-            if (it.attributes.getNamedItem("SessionIndex") == null) {
-                throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_j,
-                        message = "Single Logout support found in IdP metadata, but no " +
-                                "SessionIndex was found.",
-                        node = response)
-            }
-        }
-    }
-
-    private fun verifyAssertionsHaveAuthnStmts(assertion: Node) {
-        if (assertion.children("AuthnStatement").isEmpty()) {
-            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_i,
-                    message = "A bearer Assertion was found without an AuthnStatement.",
-                    node = response)
-        }
-    }
-
-    @Suppress("ComplexCondition")
-    private fun verifyBearerSubjectConfirmations(assertion: Node) {
-        val bearerSubjectConfirmations = assertion.children(SUBJECT)[0]
-                .children("SubjectConfirmation")
-                .filter {
-                    it.attributes.getNamedItem("Method").textContent ==
-                            "urn:oasis:names:tc:SAML:2.0:cm:bearer"
-                }.toList()
-
-        //todo - We can't throw an exception here because the spec says there could be assertions
-        // without bearer SubjectConfirmation
-        if (bearerSubjectConfirmations.isEmpty())
+        assertions.forEach { verifyIssuer() }
+
+        val bearerSubjectConfirmations = mutableListOf<Node>()
```

Comment:
✏️ Instead of mutable lists and this two-step processing to fill them both, you can do this as a single operation:
```kotlin
        val (bearerSubjectConfirmations, bearerAssertions) = assertions
                .filter { it.children(SUBJECT).isNotEmpty() }
                .flatMap { it.children(SUBJECT) }
                .filter { it.children("SubjectConfirmation").isNotEmpty() }
                .flatMap { it.children("SubjectConfirmation") }
                .filter { it.attributes.getNamedItem("Method").textContent == BEARER }
                .map { it to it.parentNode.parentNode }
                .unzip()
```